src/main/java/xfacthd/framedblocks/common/compat/jei/camo/CamoCraftingHelper.java

```diff
@@ -0,0 +1,251 @@
+    {
+        return List.of(
+                getEmptyFrameStacks(recipe.getFrameStacks()),
+                Arrays.asList(recipe.getCopyTool().getItems()),
+                // pick a prime number count so that more combinations are shown over time
+                getCamoExampleStacks(recipe.getFirstIngredient(), 97),
+                // pick a lower number so that the blank ingredient is shown more often
+                getDoubleCamoExampleStacks(recipe.getSecondIngredient(), 11)
+        );
+    }
+
+    private List<ItemStack> getCamoExampleStacks(Ingredient ingredient, int count)
+    {
+        if (ingredient.equals(camoExamplesIngredient))
+        {
+            Collections.shuffle(this.camoExamples);
+            if (count < this.camoExamples.size())
+            {
+                return new ArrayList<>(this.camoExamples.subList(0, count));
+            }
+            return new ArrayList<>(this.camoExamples);
+        }
+
+        return Arrays.asList(ingredient.getItems());
+    }
+
+    private List<ItemStack> getDoubleCamoExampleStacks(Ingredient ingredient, int count)
+    {
+        if (ingredient.equals(camoExamplesIngredient))
+        {
+            Collections.shuffle(this.camoExamples);
+            List<ItemStack> results = new ArrayList<>();
+
+            results.add(ItemStack.EMPTY);
+            count--;
+
+            if (count < this.camoExamples.size())
+            {
+                results.addAll(this.camoExamples.subList(0, count));
+            }
+            else
+            {
+                results.addAll(this.camoExamples);
+            }
+            return results;
+        }
+
+        return Arrays.asList(ingredient.getItems());
+    }
+
+    private List<ItemStack> getEmptyFrameStacks(Ingredient ingredient)
+    {
+        if (ingredient.equals(emptyFramesIngredient))
+        {
+            return emptyFramedBlocks;
+        }
+        else if (ingredient.equals(emptyDoubleFramesIngredient))
+        {
+            return emptyDoubleFramedBlocks;
+        }
+
+        return Arrays.asList(ingredient.getItems());
+    }
+
+    public void setRecipe(JeiCamoApplicationRecipe recipe, IRecipeLayoutBuilder builder, ICraftingGridHelper craftingGridHelper)
+    {
+        List<Pair<String, List<ItemStack>>> namedInputs = List.of(
+                Pair.of("frames", getEmptyFrameStacks(recipe.getFrameStacks())),
+                Pair.of("copyTool", Arrays.asList(recipe.getCopyTool().getItems())),
+                // pick a prime number count so that more combinations are shown over time
+                Pair.of("camoOne", getCamoExampleStacks(recipe.getFirstIngredient(), 97)),
+                // pick a lower number so that the blank ingredient is shown more often
+                Pair.of("camoTwo", getDoubleCamoExampleStacks(recipe.getSecondIngredient(), 11))
+        );
+        List<List<ItemStack>> inputs = getDisplayInputs(recipe);
```

Comment:
I'm from the era of java and minecraft where crash reports only gave you a line number, so I adopted the style of only having one method call per line. That way it was easier to identify what was crashing, instead of wondering which thing on the line had an NPE or whatever.
Thankfully we've moved on from that but I still find code more clear with one thing happening per line now.

Happy to adapt to your style for your project, just figured I'd explain since you wanted to know!